rides-and-fares/src/main/java/org/apache/flink/training/exercises/ridesandfares/RidesAndFaresExercise.java

```diff
@@ -18,68 +18,122 @@
+import org.apache.flink.api.common.JobExecutionResult;
 import org.apache.flink.configuration.Configuration;
 import org.apache.flink.streaming.api.datastream.DataStream;
 import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
 import org.apache.flink.streaming.api.functions.co.RichCoFlatMapFunction;
+import org.apache.flink.streaming.api.functions.sink.PrintSinkFunction;
+import org.apache.flink.streaming.api.functions.sink.SinkFunction;
+import org.apache.flink.streaming.api.functions.source.SourceFunction;
+import org.apache.flink.training.exercises.common.datatypes.RideAndFare;
 import org.apache.flink.training.exercises.common.datatypes.TaxiFare;
 import org.apache.flink.training.exercises.common.datatypes.TaxiRide;
 import org.apache.flink.training.exercises.common.sources.TaxiFareGenerator;
 import org.apache.flink.training.exercises.common.sources.TaxiRideGenerator;
-import org.apache.flink.training.exercises.common.utils.ExerciseBase;
 import org.apache.flink.training.exercises.common.utils.MissingSolutionException;
 import org.apache.flink.util.Collector;
 
 /**
- * The "Stateful Enrichment" exercise of the Flink training in the docs.
+ * The Stateful Enrichment exercise from the Flink training.
  *
  * <p>The goal for this exercise is to enrich TaxiRides with fare information.
  */
-public class RidesAndFaresExercise extends ExerciseBase {
+public class RidesAndFaresExercise {
+
+    private final SourceFunction<TaxiRide> rideSource;
+    private final SourceFunction<TaxiFare> fareSource;
+    private final SinkFunction<RideAndFare> sink;
+
+    /** Creates a job using the sources and sink provided. */
+    public RidesAndFaresExercise(
+            SourceFunction<TaxiRide> rideSource,
+            SourceFunction<TaxiFare> fareSource,
+            SinkFunction<RideAndFare> sink) {
+
+        this.rideSource = rideSource;
+        this.fareSource = fareSource;
+        this.sink = sink;
+    }
 
     /**
-     * Main method.
+     * Creates and executes the pipeline using the StreamExecutionEnvironment provided.
      *
      * @throws Exception which occurs during job execution.
+     * @param env The {StreamExecutionEnvironment}.
+     * @return {JobExecutionResult}
      */
-    public static void main(String[] args) throws Exception {
-
-        // set up streaming execution environment
-        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
-        env.setParallelism(ExerciseBase.parallelism);
+    public JobExecutionResult execute(StreamExecutionEnvironment env) throws Exception {
 
+        // A stream of taxi ride START events, keyed by rideId.
         DataStream<TaxiRide> rides =
-                env.addSource(rideSourceOrTest(new TaxiRideGenerator()))
+                env.addSource(rideSource)
                         .filter((TaxiRide ride) -> ride.isStart)
                         .keyBy((TaxiRide ride) -> ride.rideId);
```

Comment:
I think, we changed these lambdas to their simpler form in the LongRides exercises, but for educational purposes, they could also stay as they are...
```suggestion
                        .filter(ride -> ride.isStart)
                        .keyBy(ride -> ride.rideId);
```

Reply:
(this code change also requires a new spotlessApply)